src/groups/NewGroupModal.test.tsx

```diff
@@ -0,0 +1,135 @@
+      <NewGroupModal onDismiss={jest.fn()} onSuccess={jest.fn()} existingGroups={[]} />
+    );
+    // Act
+    const nameInput = getByLabelText('Enter a unique name *');
+    fireEvent.change(nameInput, { target: { value: 'Invalid Name&' } });
+
+    // Assert
+    waitFor(() =>
+      expect(getByText('Group name can only contain letters, numbers, underscores, and dashes')).toBeInTheDocument()
+    );
+  });
+
+  it('detects when the group name is empty but has been changed', async () => {
+    // Arrange
+    const user = userEvent.setup();
+    const { getByText, getByLabelText } = render(
+      <NewGroupModal onDismiss={jest.fn()} onSuccess={jest.fn()} existingGroups={[]} />
+    );
+    // Act
+    const nameInput = getByLabelText('Enter a unique name *');
+    fireEvent.change(nameInput, { target: { value: 'Valid Name' } });
+    await waitFor(() => expect(nameInput).toHaveValue('Valid Name'));
+    await user.clear(nameInput);
+    // Assert
+    expect(getByText("Group name can't be blank")).toBeInTheDocument();
+  });
+
+  it('does not allow a group name that already exists ', async () => {
+    // Arrange
+    const existingName = 'Existing name';
+    const { getByText, getByLabelText } = render(
+      <NewGroupModal onDismiss={jest.fn()} onSuccess={jest.fn()} existingGroups={[existingName]} />
+    );
+    // Act
+    const nameInput = getByLabelText('Enter a unique name *');
+    fireEvent.change(nameInput, { target: { value: existingName } });
+    // Assert
+    await waitFor(() => expect(getByText('Group name already exists')).toBeInTheDocument());
+  });
+
+  it('calls submit function on form submission with valid data', async () => {
+    // Arrange
+    const user = userEvent.setup();
+    const mockCreateFn = jest.fn().mockReturnValue(Promise.resolve());
+    const mockSetPolicyFn = jest.fn().mockReturnValue(Promise.resolve());
+    asMockedFn(Ajax).mockImplementation(
+      () =>
+        ({
+          Groups: {
+            group: jest.fn().mockImplementation(() => ({
+              create: mockCreateFn,
+              setPolicy: mockSetPolicyFn,
+            })),
+          },
+        } as DeepPartial<AjaxContract> as AjaxContract)
+    );
+    const mockOnSuccessFn = jest.fn();
+    const { getByText, getByLabelText } = render(
+      <NewGroupModal onDismiss={jest.fn()} onSuccess={mockOnSuccessFn} existingGroups={[]} />
+    );
+    const nameInput = getByLabelText('Enter a unique name *');
+    fireEvent.change(nameInput, { target: { value: 'ValidName' } });
+    await waitFor(() => expect(nameInput).toHaveValue('ValidName'));
+    const submitButton = getByText('Create Group');
+    expect(submitButton).toBeEnabled();
+    // Act
+    await user.click(submitButton);
+    // Assert
+    await waitFor(() => expect(mockCreateFn).toHaveBeenCalled());
+    await waitFor(() => expect(mockSetPolicyFn).toHaveBeenCalledWith('admin-notifier', true));
+    await waitFor(() => expect(mockCreateFn).toHaveBeenCalled());
```

Comment:
You have the `mockCreateFn` check twice (and I don't think you need all the waitFors).

Reply:
You're right - after the `fireEvent` they're needed, but not here.